hibernate-reactive-core/src/main/java/org/hibernate/reactive/context/impl/VertxContext.java

```diff
@@ -36,12 +40,21 @@ public void injectServices(ServiceRegistryImplementor serviceRegistry) {
 
     @Override
     public <T> T get(Class<T> key, String id) {
-        return getOrCreateContext().getLocal( id );
+        final io.vertx.core.Context context = currentContext();
+        if ( context != null ) {
+            return context.getLocal( id );
+        }
+        else {
+            return null;
+        }
```

Comment:
Are you sure about this? Isn't it useful to create a context for them if there is none active?

Reply:
Well not creating a Context when one isn't required is precisely the point of my performance fix.

Besides, with a method called like this I wouldn't expect to use it to create a context if there isn't one? Would people want to rely on such side-effects?

Reply:
But then this is a change to functionality that we had already discussed and agreed on a while ago. We had said that we would do this for you if you called the `Session` outside of a Vert.x context.

I mean, I'm perfectly happy to re-discuss it...

Reply:
@gavinking But why do we need to create the context here? Isn't it created anyway in the `execute`?

Reply:
This is called from `withSession()`.

Reply:
Well, actually perhaps there's room for some cleanup here: `newSession()` _does_ call `execute()` but that's just slightly too late. We should look a little more carefully at that.

Reply:
Anyway, if the idea is that calling `getOrCreateContext()` is a performance bug, then we should be getting rid of it from all the `Session` methods that call it (which is .... _all of them_). Getting rid of it from `Context.get()` is pointless, since that's called only once per session.

Reply:
I'm not arguing about where it's used: if opening a Session needs to create a context, it can of course do so.

 Just in terms of API contract of `Context#get` it seems not its responsibility to create a Context automagically - and we can save some cycles in doing so.

Reply:
We might also want to update the javadoc of `Context` or `VertxContext`. it seems we don't clarify when we expect  a new context to be created.

Reply:
Oh, wait, @Sanne, I think I understand what you're saying now: you're saying it doesn't need to be created in `get()` because it will be created later in `put()`. 

OK, sure, fine. Indeed, since `newSession()` calls `execute()`, I guess we don't even need to call `getOrCreateContext()` from `put()`, since it will have already been called from `execute()`.

Reply:
Right :)